jcl/src/java.base/share/classes/openj9/internal/tools/attach/target/AttachHandler.java

```diff
@@ -551,16 +645,36 @@ public void run() {
 					int timeout = 100;
 					int retryNumber = 0;
 					while (System.nanoTime() < shutdownDeadlineNs) {
-						currentAttachThread.join(timeout); /* timeout in milliseconds */
+						/* timeout in milliseconds */
+						currentAttachThread.join(timeout);
 						if (currentAttachThread.getState() != Thread.State.TERMINATED) {
-							IPC.logMessage(currentAttachThread + "Timeout waiting for wait loop termination.  Retry #" + retryNumber); //$NON-NLS-1$
+							IPC.logMessage(currentAttachThread + " timeout waiting for wait loop termination. Retry #" //$NON-NLS-1$
+									+ retryNumber + ", timeout = " + timeout); //$NON-NLS-1$
 							timeout *= 2;
 							AttachHandler.terminateWaitLoop(true, retryNumber);
-							++retryNumber;
+							retryNumber += 1;
 						} else {
 							break;
 						}
 					}
+					if ((fileAccessTimeUpdaterThread != null) && fileAccessTimeUpdaterThread.isAlive()) {
+						// keep existing retryNumber
+						int timeoutFileAccessUpdater = 100;
+						while (System.nanoTime() < shutdownDeadlineNs) {
```

Comment:
This should take the timeoutFileAccessUpdater time into account. The current time might be less than the shutdown deadline, but the timeoutFileAccessUpdater may have grown quite large and end of extending that time.

Reply:
Added `teardownHook.threadJoinHelper()` to adjust the `timeout` for `join()` according to the `shutdownDeadlineNs`.